src/core/infiniteQueryBehavior.ts

```diff
@@ -8,7 +8,43 @@ export function infiniteQueryBehavior<
   TData
 >(): QueryBehavior<TQueryFnData, TError, InfiniteData<TData>> {
   return {
+
     onFetch: context => {
+      context.onQueryFn = () => {
+        const fetchMore = context.fetchOptions?.meta?.fetchMore
+        const pageParam = fetchMore?.pageParam
+        const oldPages = context.state.data?.pages || []
+        const isFetchingNextPage = fetchMore?.direction === 'forward'
+        const isFetchingPreviousPage = fetchMore?.direction === 'backward'
+
+        let param: unknown|undefined;
+
+        // first page
+        if (!oldPages.length) {
+          param = undefined
+        }
+        // Fetch next page?
+        else if (isFetchingNextPage) {
+          const manual = typeof pageParam !== 'undefined'
+          param = manual
+            ? pageParam
+            : getNextPageParam(context.options, oldPages)
+        }
+        // Fetch previous page?
+        else if (isFetchingPreviousPage) {
+          const manual = typeof pageParam !== 'undefined'
+          param = manual
+            ? pageParam
+            : getPreviousPageParam(context.options, oldPages)
+        }
+
+        const queryFnContext: QueryFunctionContext = {
+          queryKey: context.queryKey,
+          pageParam: param
+        }
+
+        context.options.onQuery?.(queryFnContext)
```

Comment:
- none of this needs to be done if `context.options.onQuery` is `undefined`
- i see a bit of an overlap with what we are doing in `context.fetchFn`. not sure if it's good to call `getNextPageParam` / `getPreviousPageParam` twice - once for `onQuery` and once for the actual fetch. I'm not sure here so I'd like to wait on more feedback from either @tannerlinsley or @boschni

Reply:
Hi Jeffrey! Think this can also be implemented in the `QueryObserver.onQueryUpdate` method. There you should be able to listen for the fetch action and trigger the callback. Then it is consistent with the other callbacks and it also works outside of the hooks. Regarding the naming, maybe `onFetch` would be a bit more descriptive? Then it also corresponds more with the `fetchFn` option.

Reply:
Please see previous comment, this whole code-block is unnecessary because `onQuery` is now called from `queryFn` .

Reply:
My comment was not really related to this piece of code, just placed it here because I got tagged ;)

Reply:
> maybe onFetch would be a bit more descriptive?

I like both `onQuery` as well as `onFetch` 🤔 . `onQuery` for `useQuery` corresponds well with `onMutate` for `useMutation`.`onFetch` would also make sense because we have the `isFetching` flag 🤷

Reply:
Yeah it also relates to the `refetch()` method in `useQuery`. When doing mutations you call `mutate()` so it makes sense to use `onMutate` there.

Reply:
if you prefer `onFetch`, then let's rename 👍

Reply:
Thanks guys! I'll rename to `onFetch` and move the call to `QueryObserver.onQueryUpdate`.

Reply:
@TkDodo @boschni 
Update:

- I renamed the function `onFetch`.
- The function is called from the `queryObserver`

Let me know what you think :-).

Reply:
@TkDodo @boschni 
- I renamed the function `onFetch`
- I moved the call to `QueryObserver.onQueryUpdate`

Let me know what you think :-)